### PR TITLE
fix Inline temporary variable adds unnecessary cast

### PR DESCRIPTION
**Pull request by pyoub**
trying to fix Inline temporary variable adds unnecessary cast

Fixes: #69869

### CODE REVIEW

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -414,7 +414,7 @@ ExpressionSyntax CreateExpressionToInline()
 
                     return SyntaxFactory.ArrayCreationExpression(arrayType, arrayInitializer);
                 }
-                else if (isVar && expression is ObjectCreationExpressionSyntax or ArrayCreationExpressionSyntax or CastExpressionSyntax)
+                else if (isVar && expression is ObjectCreationExpressionSyntax or ArrayCreationExpressionSyntax or CastExpressionSyntax or InvocationExpressionSyntax)
```
Comment (DoctorKrolic) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:417]: That is most likely not the right fix unfortunatelly

Comment (pyoub) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:417]: so what you suggest?

File: src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs [hunk 2]
```diff
@@ -22,7 +22,6 @@
 using Microsoft.CodeAnalysis.Editor.UnitTests.Diagnostics;
 using Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces;
 using Microsoft.CodeAnalysis.Host;
-using Microsoft.CodeAnalysis.Remote;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:25]: undo this.

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:417]: i agree with DoctorKrolic.  thsi cahnge doesn't really make sense.  the comment below this explains the rational for why we don't need to add the cast.  however, for an invocatino expression we still *do* need to cast to preserve semantics in many cases.

Comment (pyoub) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:417]: @CyrusNajmabadi  can i have an example when we need to do the cast on InvocationExpressionSyntax ?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:417]: Ah, i think i missed that this iwas the `is var` case.  In that case, it's likely safe to whitelist invocations as well.  Can you update the comment inside to make this clearer?  Thanks!

Comment (pyoub) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:417]: done!

File: src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs (hunk 2 above)
Comment (pyoub) [reply src/EditorFeatures/DiagnosticsTestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs]: done!

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 3]
```diff
@@ -414,8 +414,9 @@ ExpressionSyntax CreateExpressionToInline()
...
                 {
+                    // if we have an InvocationExpressionSyntax `var x = Math.Round(1.1D);` there's no need to do any casting.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:419]: needs an explanation why this is.

Comment (pyoub) [reply src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs]: is that clear ?

File: src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs [hunk 4]
```diff
@@ -414,8 +414,9 @@ ExpressionSyntax CreateExpressionToInline()
 
                     return SyntaxFactory.ArrayCreationExpression(arrayType, arrayInitializer);
                 }
-                else if (isVar && expression is ObjectCreationExpressionSyntax or ArrayCreationExpressionSyntax or CastExpressionSyntax)
+                else if (isVar && expression is ObjectCreationExpressionSyntax or ArrayCreationExpressionSyntax or CastExpressionSyntax or InvocationExpressionSyntax)
                 {
+                    // if we have an InvocationExpressionSyntax `var x = Math.Round(1.1D);` there's no need to do any casting as x is var and the method return type is known.
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/InlineTemporary/InlineTemporaryCodeRefactoringProvider.cs:419]:
```suggestion
                    // if we have an InvocationExpressionSyntax `var x = Math.Round(1.1D);` there's no need to do any casting
                    // as x's type will be entirely determined by the return type of the invoked expression.
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by pyoub** [comment]
i don't know which tests are failing

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Pull request merged**